Cowbuilder mirror and keyring selection

yvlasenko opened the pull request:
Added cowbuilder mirror and keyring selection based on suite name. Without it I could not build debian base on ubuntu, maybe there is another way. Tested on trusty host to build wheezy/sid bases.

Needs debian-archive-keyring on ubuntu and ubuntu-archive-keyring on debian.

mika commented:
Hm, what I usually do to e.g. support building Ubuntu on Debian with custom options is something like:

    export COWBUILDER_DIST=$distribution
    /usr/bin/build-and-provide-package

in the build step, then inside /etc/jenkins/debian_glue set:

    PBUILDER_CONFIG=/etc/jenkins/pbuilderrc

and finally adjust /etc/jenkins/pbuilderrc, like:

    if [ "$distribution" = "precise" ] ; then
      MIRRORSITE="http://ubuntu.inode.at/ubuntu/"
      # we need key id 40976EAF437D05B5
      DEBOOTSTRAPOPTS=("${DEBOOTSTRAPOPTS[@]}" "--keyring=/usr/share/keyrings/ubuntu-archive-keyring.gpg")
      # cowdancer is in universe
      COMPONENTS="main universe"
    else
      MIRRORSITE="http://debian.example.local/debian"
    fi

Something like that in the reverse direction should work for you too, right?

Note: I hereby don't say that your PR is unneeded, I'm just afraid of setting defaults where we might break setups like the above one.

yvlasenko commented:
Good point, I didn't think of customizing pbuilderrc. I'll give it a try, thanks!

yvlasenko commented:
Your solution works like a charm!

My pbuilderrc:
```
DEBIAN_SUITES=("wheezy" "jessie" "sid")
DEBIAN_MIRROR="ftp.us.debian.org"
UBUNTU_MIRROR="mirrors.kernel.org"

if $(echo ${DEBIAN_SUITES[@]} | grep -q "${DIST}"); then
    # Debian configuration
    MIRRORSITE="http://$DEBIAN_MIRROR/debian/"
    COMPONENTS="main contrib non-free"
    DEBOOTSTRAPOPTS=("${DEBOOTSTRAPOPTS[@]}" "--keyring=/usr/share/keyrings/debian-archive-keyring.gpg")
    OTHERMIRROR="deb [trusted=yes] http://localhost/debian ${DIST}/"
    if [ "${DIST}" = "wheezy" ] ; then
        OTHERMIRROR="$OTHERMIRROR | deb $MIRRORSITE wheezy-backports $COMPONENTS"
    fi
else
    # Ubuntu configuration
    MIRRORSITE="http://$UBUNTU_MIRROR/ubuntu/"
    COMPONENTS="main restricted universe multiverse"
    DEBOOTSTRAPOPTS=("${DEBOOTSTRAPOPTS[@]}" "--keyring=/usr/share/keyrings/ubuntu-archive-keyring.gpg")
    OTHERMIRROR="deb [trusted=yes] http://localhost/ubuntu ${DIST}/"
fi
```

I think you should mention keyring issue and workaround in FAQ.

yvlasenko commented:
Removed cowbuilder mirror and keyring selection and removed repository cleaning for dput as it doesn't work and break things when no reprepro repository found.

mika commented:
The sed issue is supported via ADJUST_DISTRIBUTION_ONTHEFLY now: https://github.com/mika/jenkins-debian-glue/commit/d9cc2bcca2386f6db97460c5ef74a589f0341017

Thanks for your PR and feedback!
